test/celeritas/user/Diagnostic.test.cc

```diff
@@ -0,0 +1,144 @@
+#include "celeritas/global/Stepper.hh"
+#include "celeritas/global/alongstep/AlongStepUniformMscAction.hh"
+#include "celeritas/phys/PDGNumber.hh"
+#include "celeritas/phys/ParticleParams.hh"
+#include "celeritas/phys/Primary.hh"
+#include "celeritas/user/ActionDiagnostic.hh"
+
+#include "../TestEm3Base.hh"
+#include "DiagnosticTestBase.hh"
+#include "celeritas_test.hh"
+
+using celeritas::units::MevEnergy;
+
+namespace celeritas
+{
+namespace test
+{
+//---------------------------------------------------------------------------//
+// TEST FIXTURES
+//---------------------------------------------------------------------------//
+
+#define TestEm3DiagnosticTest TEST_IF_CELERITAS_GEANT(TestEm3DiagnosticTest)
+class TestEm3DiagnosticTest : public TestEm3Base, public DiagnosticTestBase
+{
+    SPConstAction build_along_step() override
+    {
+        auto& action_reg = *this->action_reg();
+        UniformFieldParams field_params;
+        field_params.field = {0, 0, 1 * units::tesla};
+        auto msc = UrbanMscParams::from_import(
+            *this->particle(), *this->material(), this->imported_data());
+
+        auto result = std::make_shared<AlongStepUniformMscAction>(
+            action_reg.next_id(), field_params, msc);
+        CELER_ASSERT(result);
+        CELER_ASSERT(result->has_msc());
+        action_reg.insert(result);
+        return result;
+    }
+
+    VecPrimary make_primaries(size_type count) override
+    {
+        Primary p;
+        p.energy = MevEnergy{10.0};
+        p.position = {-22, 0, 0};
+        p.direction = {1, 0, 0};
+        p.time = 0;
+        std::vector<Primary> result(count, p);
+
+        auto electron = this->particle()->find(pdg::electron());
+        CELER_ASSERT(electron);
+        auto positron = this->particle()->find(pdg::positron());
+        CELER_ASSERT(positron);
+
+        for (auto i : range(count))
+        {
+            result[i].event_id = EventId{0};
+            result[i].track_id = TrackId{i};
+            result[i].particle_id = (i % 2 == 0 ? electron : positron);
+        }
+        return result;
+    }
+};
+
+//---------------------------------------------------------------------------//
+// TESTEM3
+//---------------------------------------------------------------------------//
+
+TEST_F(TestEm3DiagnosticTest, host)
+{
+    auto result = this->run<MemSpace::host>(256, 32);
+
+    // Check action diagnostic results
+    if (CELERITAS_CORE_GEO == CELERITAS_CORE_GEO_VECGEOM)
+    {
+        static const size_type expected_action_counts[] = {
+            0u,  0u, 0u, 0u, 0u,  0u,   0u,    1580u, 1168u, 0u,   1019u, 55u,
+            0u,  0u, 0u, 0u, 22u, 90u,  283u,  0u,    0u,    575u, 0u,    0u,
+            19u, 0u, 0u, 0u, 0u,  127u, 0u,    20u,   14u,   0u,   467u,  386u,
+            0u,  0u, 0u, 0u, 0u,  0u,   1769u, 289u,  284u,  0u,   15u,   10u,
+            0u,  0u, 0u, 0u, 0u,  0u,   0u,    0u,    0u};
+        EXPECT_VEC_EQ(expected_action_counts, result.action_counts);
```

Comment:
Since these results will undoubtedly change with variations on Geant4 and data, we probably need something like the `is_ci()` testing from Stepper... maybe also another way of testing this would be to make a list of indices of the nonzero counts and make sure those match?